source/d3d11/d3d11_device_context.cpp

```diff
@@ -16,23 +16,254 @@ void D3D11DeviceContext::log_drawcall(UINT vertices)
+{	
+	screen_dimensions();
+	float aspect_ratio = ((float)_screen_width) / ((float)_screen_height);
+	float twfactor = 1.0f;
+	float thfactor = 1.0f;
+
+	// check the depthstencil flag
+	if (texture_desc.BindFlags != (D3D11_BIND_DEPTH_STENCIL | D3D11_BIND_SHADER_RESOURCE) && texture_desc.BindFlags != D3D11_BIND_DEPTH_STENCIL)
+	{
+		return false;
+	}
+
+	if (texture_desc.Width != _screen_width)
+	{
+		twfactor = (float)_screen_width / texture_desc.Width;
+	}
+	if (texture_desc.Height != _screen_height)
+	{
+		thfactor = (float)_screen_height / texture_desc.Height;
+	}
+
+	// check if aspect ratio is similar to the back buffer one
+	float stencilaspectratio = ((float)texture_desc.Width) / ((float)texture_desc.Height);
+	if (fabs(stencilaspectratio - aspect_ratio) > 0.1f)
+	{
+		return false;
+	}
+
+	return true;
+}
+
 void D3D11DeviceContext::set_active_depthstencil(ID3D11DepthStencilView* pDepthStencilView)
 {
-	if (pDepthStencilView != _active_depthstencil)
+	if (pDepthStencilView != nullptr)
+	{
+		if (pDepthStencilView != _active_depthstencil)
+		{
+			_active_depthstencil = pDepthStencilView;
+			_draw_call_tracker.track_depthstencil(_active_depthstencil.get());
+		}
+	}
+}
+
+void D3D11DeviceContext::set_active_OM_depthstencil(ID3D11DepthStencilView* pDepthStencilView)
+{
+	if (pDepthStencilView != nullptr)
+	{
+		com_ptr<ID3D11Texture2D> texture;
+		com_ptr<ID3D11Resource> resource;
+		D3D11_TEXTURE2D_DESC texture_desc;
+
+		pDepthStencilView->GetResource(&resource);
+		if (!FAILED(resource->QueryInterface(&texture)))
+		{
+			texture->GetDesc(&texture_desc);
+		}
+		else
+		{
+			return;
+		}
+
+		if (!check_depthstencil(texture, texture_desc))
+		{
+			return;
+		}
+
+		if (_OM_iter >= _clear_DSV_iter)
+		{
+			set_active_depthstencil(pDepthStencilView);
+			// copy the depth stencil texture
+			_depth_texture = texture.get();
+
+			if (_depth_texture != nullptr)
+			{
+				_draw_call_tracker.set_depth_texture(_depth_texture.get());
+			}
+		}
+	}
+}
+
+void D3D11DeviceContext::set_active_cleared_depthstencil(ID3D11DepthStencilView* pDepthStencilView)
+{
+	if (pDepthStencilView != nullptr)
 	{
-		_active_depthstencil = pDepthStencilView;
+		com_ptr<ID3D11Texture2D> texture;
+		com_ptr<ID3D11Resource> resource;
+		D3D11_TEXTURE2D_DESC texture_desc;
 
-		if (pDepthStencilView != nullptr)
+		pDepthStencilView->GetResource(&resource);
+		if (!FAILED(resource->QueryInterface(&texture)))
 		{
-			_draw_call_tracker.track_depthstencil(pDepthStencilView);
+			texture->GetDesc(&texture_desc);
+		}
+		else
+		{
+			return;
+		}
+
+		if (!check_depthstencil(texture, texture_desc))
+		{
+			return;
+		}
+
+		reshade::d3d11::draw_call_tracker::depthstencil_counter_info counters = _draw_call_tracker.get_counters();
+
+		if (reshade::runtime::depth_buffer_clearing_number == 0)
+		{
+			set_active_depthstencil(pDepthStencilView);
+			// copy the depth stencil texture
+			_depth_texture = copy_depth_texture(texture_desc, texture.get());
```

Comment:
This looks like a bad idea. Creating a new texture every frame is a sure way to kill performance. Why not use the `_depthstencil_texture` field in `d3d11_runtime` and keep a persistent texture around that you can copy into? I feel like large parts of this algorithm should be moved to the `d3d11_runtime` class anyway ... I'm still not a fan of the `static` variables used here, which doing so would solve as well.

Reply:
That is what i thought at first, but the copyResource method seems to be an asynchronous one, so it does not kill the performance a lot, especially is you call it only once per frame. This method also copy from GPU VRAM to GPU VRAM.
 In my tests, i did not encounter big losses in perfs.
Keep a persistent texture implies that you make a copy of the texture before it is cleared by the game engine, so you still need a copyResource method. Another way to do it, would be to unlink the depth texture to the depthstencil before the engine clears this texture then, in the final rendering stage, relink this texture to the depthstencil, but i do not see a good way to do so....

Reply:
I'm not concerned about the `CopyResource`. That one is fast.I'm concerned about the `CreateTexture`. That one should not be called every frame. Allocation is slow.

Reply:
Ok, i understand what you mean. Thks for having comitted my pull request ;-)